### PR TITLE
Preparation for Hyperproperties - part 2 (strategies)

### PR DESCRIPTION
**Pull request by h0nzZik**
I would also like to run `./t/propagate-*` tests in CI. @nishantjr any idea how to do that?

### CODE REVIEW

File: prover/lang/kore-lang.md [hunk 1]
```diff
@@ -472,6 +472,25 @@ and values, passed to K's substitute.
   rule getLength(.Patterns) => 0
   rule getLength(P, Ps) => 1 +Int getLength(Ps)
 
+  syntax Pattern ::= getLast(Patterns) [function]
+  rule getLast(Ps) => getMember(getLength(Ps) -Int 1, Ps)
+
+  syntax Patterns ::= takeFirst(Int, Patterns) [function]
```
Comment (nishantjr) [inline prover/lang/kore-lang.md:478]: `head` and `tail` may be better names?

File: prover/lang/kore-lang.md [hunk 2]
```diff
@@ -778,6 +797,13 @@ Simplifications
   rule hasImplicationContext(\typeof(P, _))
     => hasImplicationContext(P)
 
+  syntax Pattern ::= "#\\exists" "{" Patterns "}" Pattern [function]
```
Comment (nishantjr) [inline prover/lang/kore-lang.md]: This needs a better name. There are already functions called `#exists`, `#not` that propagate these constructs into patterns.

File: prover/lang/kore-lang.md [hunk 3]
```diff
@@ -778,6 +797,13 @@ Simplifications
...
+  rule #\exists{.Patterns} P => P
+  rule #\exists{V, Vs} P => \exists{V, Vs} P
+
+  syntax Pattern ::= "#\\and" "(" Patterns ")" [function]
```
Comment (nishantjr) [inline prover/lang/kore-lang.md]: Not sure about hyperlogics, but for SL, we've found it easier to standardize on a normal form where `\exists`, `\and` etc are always present in a formula irrespective of whether it is needed. The normal form we use is `\implies(\and(...), \exists {... } \and(...))`

**Review by nishantjr — CHANGES_REQUESTED** [review]
Besides the above, looks good.

File: prover/lang/kore-lang.md (hunk 1 above)
Comment (h0nzZik) [reply prover/lang/kore-lang.md:478]: `head` would return only the first item from the list, while `takeFirst` returns a list of items. Similarly, `tail` returns the list of everything except head, while `getLast` returns the last item: like `head(reverse(l))`.

**Pull request merged**